Fix eigenvects NotImplementedError for matrices with algebraic coefficients

Nikhil172913832 opened the pull request:
Fixes #28507

Replace CRootOf(expr, var, idx) with CRootOf.all_roots(poly) to preserve AlgebraicField domain and avoid conversion to unsupported EX domain.

<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->

Fixes #28507

#### Brief description of what is fixed or changed

The `eigenvects()` method was raising `NotImplementedError: CRootOf is not supported over EX` when computing eigenvectors for matrices with algebraic coefficients (e.g., sqrt(2), sqrt(3), sqrt(5), sqrt(7)).

**Root cause:** Converting the minimal polynomial to an expression via `minpoly.as_expr()` before passing to `CRootOf` caused the domain to convert from `AlgebraicField` to `EX` domain during `PurePoly` construction and subsequent `retract()` call.

**Solution:** Use `CRootOf.all_roots(poly)` with the `Poly` object directly instead of creating individual `CRootOf` objects from an expression. This preserves the `AlgebraicField` domain and ensures the correct code path (`_get_roots_alg`) is used for handling algebraic coefficients.

#### Other comments

The slow performance (~20 minutes for large matrices) mentioned in the issue is tracked separately in #28486 and requires root refinement algorithm improvements, which is beyond the scope of this fix.

All existing tests pass without regression.

#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers. Formerly, `log(-x)` incorrectly gave `-log(x)`.

* physics.units
  * Corrected a semantical error in the conversion between volt and statvolt which
    reported the volt as being larger than the statvolt.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->

* matrices
  * Fixed `NotImplementedError` in `eigenvects()` for matrices with algebraic coefficients. The method now correctly handles polynomials over algebraic fields by preserving domain information.

<!-- END RELEASE NOTES -->

Nikhil172913832 committed:
Fix eigenvects NotImplementedError for algebraic coefficients

Fixes #28507

Replace CRootOf(expr, var, idx) with CRootOf.all_roots(poly) to preserve AlgebraicField domain and avoid conversion to unsupported EX domain.

Nikhil172913832 committed:
update .mailmap

Nikhil172913832 committed:
update .mailmap

oscarbenjamin commented:
The test function should have a more meaningful name rather than `test_issue_28507`. It should be added in this PR.

Nikhil172913832 committed:
Verify eigenvects works with algebraic field polynomials

Nikhil172913832 committed:
Add test for issue #28507

Verify eigenvects works with algebraic field polynomials

Nikhil172913832 committed:
Update test to use actual matrix from issue #28507

Nikhil172913832 committed:
Remove trailing whitespace from test

Nikhil172913832 committed:
Optimize _which_roots to avoid slow EX domain evaluation

oscarbenjamin commented on sympy/polys/polytools.py:
```diff
@@ -3975,18 +3975,33 @@ def which_all_roots(f, candidates):
+        for r in candidates:
+            if r not in seen:
+                seen.add(r)
+                roots.append(r)
```
I don't think this is needed.

Nikhil172913832 committed:
Move test to appropriate location and restore Counter

Nikhil172913832 commented:
@oscarbenjamin I’ve removed the deduplication logic and restored the previous Counter-based approach. I’ve also moved the test next to test_eigenvects().

oscarbenjamin commented on sympy/matrices/tests/test_eigen.py:
```diff
@@ -225,6 +225,18 @@ def test_eigenvects():
         assert M*vec_list[0] == val*vec_list[0]
 
 
+def test_eigenvects_algebraic_field():
```
This test should be marked as `@slow` because it takes more than a couple of seconds.

Nikhil172913832 commented:
@oscarbenjamin I tried running this version, but it was extremely slow. So I replaced the symbolic equality with a numerical tolerance check using evalf and nsimplify.
This keeps the test logically equivalent (verifying Mv=λv) while making it run in a reasonable time.
What do you think about using this approach?
``` python
@slow
def test_eigenvects_algebraic_field():
    M = Matrix([
        [45, 15, 20, 18, sqrt(5)/2],
        [15, 85, 35, 22, Rational(11, 37)],
        [20, 35, 15, 40, 3*sqrt(3)/7],
        [Rational(9, 2), 2, 4, 12, sqrt(7)/5],
        [Rational(25, 13), sqrt(7)/4, 2, 1, 1]
    ])
    vecs = M.eigenvects()
    assert len(vecs) == 5
    for val, mult, vec_list in vecs:
        assert len(vec_list) == 1
        v = vec_list[0]
        diff = M*v - val*v
        assert all(abs(nsimplify(x)) < 1e-10 for x in diff.evalf().values())
```

Nikhil172913832 committed:
extend test_eigenvects_algebraic_field to verify algebraic eigenpairs with numerical tolerance

oscarbenjamin commented:
I was hoping that @vprayag2005 could review this.

vprayag2005 commented:
Did you check the performance before and after your change? I think using all_roots may slow it down a little bit. I mean some random matrix, e.g., of size 10×10.
